actionpack/lib/action_dispatch/journey/nodes/node.rb

```diff
@@ -4,6 +4,89 @@
 
 module ActionDispatch
   module Journey # :nodoc:
+    class Ast # :nodoc:
+      delegate :find_all, :left, :right, :to_s, :to_sym, :type, to: :tree
+      attr_reader :names, :path_params, :tree, :wildcard_options, :terminals
+      alias :root :tree
+
+      def initialize(tree, formatted)
+        @tree = tree
+        @path_params = []
+        @names = []
+        @symbols = []
+        @stars = []
+        @terminals = []
+        @glob = false
+        @wildcard_options = {}
+
+        visit_tree(formatted)
+      end
+
+      def requirements=(requirements)
+        # inject any regexp requirements for `star` nodes so they can be
+        # determined nullable, which requires knowing if the regex accepts an
+        # empty string.
+        (symbols + stars).each do |node|
+          re = requirements[node.to_sym]
+          node.regexp = re if re
+        end
+      end
+
+      def route=(route)
+        terminals.each { |n| n.memo = route }
+      end
+
+      def glob?
+        @glob
+      end
+
+      private
+        attr_reader :symbols, :stars
+
+        def visit_tree(formatted)
+          tree.each do |node|
+            if node.symbol?
+              path_params << node.to_sym
+              names << node.name
+              symbols << node
+            elsif node.star?
+              stars << node
+              @glob = true
+
+              if formatted != false
+                # Add a constraint for wildcard route to make it non-greedy and
+                # match the optional format part of the route by default.
+                wildcard_options[node.name.to_sym] ||= /.+?/
+              end
+            elsif node.cat?
+              alter_regex_for_custom_routes(node)
+            end
+
+            if node.terminal?
+              terminals << node
+            end
+          end
+        end
+
+        # Find all the symbol nodes that are adjacent to literal nodes and alter
+        # the regexp so that Journey will partition them into custom routes.
+        def alter_regex_for_custom_routes(node)
+          if node.left.literal? && node.right.symbol?
+            symbol = node.right
+          elsif node.left.literal? && node.right.cat? && node.right.left.symbol?
+            symbol = node.right.left
+          elsif node.left.symbol? && node.right.literal?
+            symbol = node.left
+          elsif node.left.symbol? && node.right.cat? && node.right.left.literal?
+            symbol = node.left
+          end
```

Comment:
Maybe we don't want to mess too much with the logic structure here, but I found this a bit hard to grok. I shortened it to this:

```ruby
  l, r, rl = node.left, node.right, node.right.left

  symbol =
    case
    when l.literal? && r.symbol?            then r
    when l.symbol?  && r.literal?           then l
    when r.cat? && l.literal? && rl.symbol? then rl
    when r.cat? && l.symbol? && rl.literal? then l
    end
```

And a pattern emerged. I think if we change it to something like this it becomes a little clear that there's two branches here — where we're doing the same search but at a different nesting:

```ruby
def alter_regex_for_custom_routes(node)
  symbol   = extract_symbol_paired_with_literal(node.left, node.right)
  symbol ||= extract_symbol_paired_with_literal(node.left, node.right.left) if node.right.cat?

  if symbol
    symbol.regexp = /(?:#{Regexp.union(symbol.regexp, '-')})+/
  end
end

def extract_symbol_paired_with_literal(left, right)
  case
  when left.literal? && right.symbol?  then right
  when left.symbol?  && right.literal? then left
  end
end
```

Totally fine if we don't go for this, since it's a little out of scope for the perf boosts here.

Reply:
Had another thought about this. This might be a simpler version of both:

```ruby
  left  = node.left
  right = node.right.cat? ? node.right.left : node.right

  symbol =
    case
    when left.literal? && right.symbol?  then right
    when left.symbol?  && right.literal? then left
    end
```

Reply:
I agree that's much easier to read. Readibility was a secondary motivation for this PR, so I'll include it!

Reply:
Thinking of tests like https://github.com/composerinteralia/rails/blob/journey-performance-improvements/actionpack/test/dispatch/routing_test.rb#L5003-L5057

Reply:
Opened https://github.com/rails/rails/pull/42911 to try out removing this code altogether. Seemed to work fine here at GitHub.

Reply:
Thanks for the advice, @kaspth! We managed to code golf this down to 0 lines of code!

Reply:
Love it 😂🏌️